### PR TITLE
Add GCE scale testing on kops

### PR DESCRIPTION
**Pull request by upodroid**
Part of https://github.com/kubernetes/enhancements/issues/4224
Fixes: https://github.com/kubernetes/kops/issues/16241

kops will be replacing the current kubeup 5k scale test, which will be developed here

### CODE REVIEW

**Comment by rifelpet** [comment]
I see the [cloud-controller-manager is logging:](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kops/16181/presubmit-kops-gce-scale-kubenet-using-cl2/1737554519111241728/artifacts/cluster-info/kube-system/cloud-controller-manager-8lwg4/logs.txt)

`E1220 20:26:59.428334       1 route_controller.go:329] Could not create route 210655d5-c5e4-4b69-9cc3-7a7e060502e6 100.97.198.0/24 for node nodes-us-east1-c-hdrg: googleapi: Error 403: QUOTA_EXCEEDED - Quota 'STATIC_ROUTES_PER_NETWORK' exceeded.  Limit: 200.0 globally.`

Probably need some quotas increased in this project?

**Comment by upodroid** [comment]
> Probably need some quotas increased in this project?

Switching to gce networking to avoid that issue.

/test presubmit-kops-gce-small-scale-kubenet-using-cl2

File: upup/pkg/fi/cloudup/gce/network.go [hunk 1]
```diff
@@ -162,17 +162,17 @@ func performNetworkAssignmentsIPAliases(ctx context.Context, c *kops.Cluster, cl
 	// CIDRs should be in the RFC1918 range, but otherwise we have no constraints
 	networkCIDR := "10.0.0.0/8"
 
-	podCIDR, err := used.Allocate(networkCIDR, net.CIDRMask(14, 32))
```
Comment (upodroid) [inline upup/pkg/fi/cloudup/gce/network.go:165]: The changes in this file supports clusters with: 
- 8k nodes
- 65k services

https://cloud.google.com/kubernetes-engine/docs/concepts/alias-ips#defaults_limits

We should create a cilium-gce network provider to resolve https://github.com/kubernetes/kops/issues/16071 and avoid creating an unreasonable number of routes per node.

File: tests/e2e/scenarios/scalability/run-test.sh [hunk 2]
```diff
@@ -141,42 +127,54 @@ KUBETEST2_ARGS+=("--kops-binary-path=${KOPS_BIN}")
...
+    --project=k8s-infra-e2e-scale-project
+  cat /etc/ssh-key-secret/ssh-public
+  mkdir -p "/tmp/kops/${CLUSTER_NAME}"
+  cp /etc/ssh-key-secret/ssh-private "/tmp/kops/${CLUSTER_NAME}/id_rsa"
+  export KUBE_SSH_KEY_PATH="/tmp/kops/${CLUSTER_NAME}/id_rsa"
+fi
+
 # More time for bigger clusters
 KUBETEST2_ARGS+=("--validation-wait=55m")
+KUBETEST2_ARGS+=("--validation-count=5") # reduce Cloud API calls
```
Comment (hakman) [inline tests/e2e/scenarios/scalability/run-test.sh:147]:
```suggestion
KUBETEST2_ARGS+=("--validation-count=3") # reduce Cloud API calls
```

Comment (upodroid) [reply tests/e2e/scenarios/scalability/run-test.sh:147]: ack

I want to run the entire thing via a single kubetest2 commands so we can log the build, cluster up and down, timings for cloud/kops regression

It seems cl2 is picky about something

File: tests/e2e/scenarios/scalability/run-test.sh [hunk 3]
```diff
@@ -86,12 +63,24 @@ echo "ADMIN_ACCESS=${ADMIN_ACCESS}"
...
+if [[ "${CLOUD_PROVIDER}" == "aws" ]]; then
+  create_args+=("--network-cidr=10.0.0.0/16,10.1.0.0/16,10.2.0.0/16,10.3.0.0/16,10.4.0.0/16,10.5.0.0/16,10.6.0.0/16,10.7.0.0/16,10.8.0.0/16,10.9.0.0/16,10.10.0.0/16,10.11.0.0/16,10.12.0.0/16")
+  create_args+=("--node-size=t3a.medium,t3.medium,t2.medium,t3a.large,c5a.large,t3.large,c5.large,m5a.large,m6a.large,m5.large,c4.large,c7a.large,r5a.large,r6a.large,m7a.large")
+  create_args+=("--node-volume-size=20")
+  create_args+=("--zones=us-east-2a,us-east-2b,us-east-2c")
+  create_args+=("--image=${INSTANCE_IMAGE:-ssm:/aws/service/canonical/ubuntu/server/20.04/stable/current/amd64/hvm/ebs-gp2/ami-id}")
+fi
+if [[ "${CLOUD_PROVIDER}" == "gce" ]]; then
+  create_args+=("--zones=us-east1-b,us-east1-c,us-east1-d")
+  create_args+=("--topology=private")
```
Comment (upodroid) [inline tests/e2e/scenarios/scalability/run-test.sh:75]: If we want to use private topology, we need to either:
- create a new topology called `private-nodes` which creates nodes with internal IPs only and control planes with both external and internal IPs
- create the cluster with `--bastion=e2e-bastion` flag, set the `KUBE_SSH_BASTION` env variable and tweak cl2 to use that to connect to the control plane instances. the e2e framework supports this https://github.com/kubernetes/kubernetes/blob/735804dc812ce647f8c130dced45b5ba4079b76e/test/e2e/framework/ssh/ssh.go#L53

**Comment by upodroid** [comment]
/test presubmit-kops-gce-scale-calico-using-cl2

here goes our first proper 5k run

File: upup/pkg/fi/cloudup/gce/op.go [hunk 4]
```diff
@@ -58,7 +58,7 @@ func waitForZoneOp(client *compute.Service, op *compute.Operation) error {
 	}
 
 	return waitForOp(op, func(operationName string) (*compute.Operation, error) {
-		return client.ZoneOperations.Get(u.Project, u.Zone, operationName).Do()
+		return client.ZoneOperations.Wait(u.Project, u.Zone, operationName).Do()
```
Comment (upodroid) [inline upup/pkg/fi/cloudup/gce/op.go:61]: WAIT is better than GET

https://cloud.google.com/compute/docs/reference/rest/v1/globalOperations/wait

**Comment by marseel** [comment]
I tried to debug https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/kops/16181/presubmit-kops-gce-scale-calico-using-cl2/1740147021265768448/

Would be great if we could set `--experimental-prometheus-snapshot-to-report-dir=true` for CL2 so we get prometheus snapshot to analyze the run and `--v=2` to get more logs.
Also, I couldn't find kube-apisever logs for more than 2-3 minutes, also we usually run apiserver with `v=3` to be able to see single API calls in logs :cry: 

From CL2 perspective it seems like some daemonsets were disappearing when CL2 was waiting for them after creation :thinking: :thinking: :thinking: 
```
(around 50 logs like this)
116 pods disappeared: daemonset-0-j2m5q, daemonset-0-nr64b, ....
```
This might have generated some additional load. 
Also, potentially apiserver had some issues based on these logs:
```
W1228 00:51:30.379466   51558 reflector.go:441] pkg/mod/k8s.io/client-go@v0.22.15/tools/cache/reflector.go:167: watch of *v1.Pod ended with: very short watch: pkg/mod/k8s.io/client-go@v0.22.15/tools/cache/reflector.go:167: Unexpected watch close - watch lasted less than a second and no items received
W1228 00:51:30.384225   51558 reflector.go:441] pkg/mod/k8s.io/client-go@v0.22.15/tools/cache/reflector.go:167: watch of <unspecified> ended with: very short watch: pkg/mod/k8s.io/client-go@v0.22.15/tools/cache/reflector.go:167: Unexpected watch close - watch lasted less than a second and no items received
```
but I don't think I have enough logs/prometheus data to debug it further.

File: tests/e2e/kubetest2-kops/deployer/dumplogs.go [hunk 5]
```diff
@@ -43,6 +44,17 @@ func (d *deployer) DumpClusterLogs() error {
...
 	}
+	version, err := kops.GetVersion(d.KopsBinaryPath)
+	if err != nil {
+		return err
+	}
+	if version > "1.29" {
+		// we don't want to dump from more than 10 nodes
+		// kops has logic to dump the control plane instances first
+		// remove this code once kops 1.28 is EoL
+		args = append(args, "--max-nodes", "10")
```
Comment (upodroid) [inline tests/e2e/kubetest2-kops/deployer/dumplogs.go:55]: It takes 30 minutes to dump logs from 500 nodes. We only need logs from the control plane and few instances.

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kops/16181/presubmit-kops-gce-small-scale-calico-using-cl2/1740355086892142592

**Comment by upodroid** [comment]
logs are super huge :( 

```
ubuntu@dev:/tmp/logs/artifacts/control-plane-us-east1-b-b1hb$ ls -alh
total 16G
drwxr-xr-x   2 ubuntu ubuntu  4.0K Dec 28 21:10 .
drwxr-xr-x 536 ubuntu ubuntu   20K Dec 28 22:04 ..
-rw-rw-r--   1 ubuntu ubuntu  8.0K Dec 28 21:10 cloud-init-output.log
-rw-rw-r--   1 ubuntu ubuntu  123K Dec 28 20:56 containerd.log
-rw-rw-r--   1 ubuntu ubuntu    45 Dec 28 21:10 dns-controller.log
-rw-rw-r--   1 ubuntu ubuntu   11M Dec 28 21:10 etcd-events.log
-rw-rw-r--   1 ubuntu ubuntu   24M Dec 28 21:10 etcd.log
-rw-rw-r--   1 ubuntu ubuntu   403 Dec 28 21:10 etchosts
-rw-rw-r--   1 ubuntu ubuntu    45 Dec 28 21:10 external-dns.log
-rw-rw-r--   1 ubuntu ubuntu  1.7K Dec 28 20:56 iptables-filter.log
-rw-rw-r--   1 ubuntu ubuntu  282K Dec 28 20:56 iptables-nat.log
-rw-rw-r--   1 ubuntu ubuntu  4.7M Dec 28 20:56 journal.log
-rw-rw-r--   1 ubuntu ubuntu   68K Dec 28 21:10 kern.log
-rw-rw-r--   1 ubuntu ubuntu  177K Dec 28 20:56 kops-configuration.log
-rw-rw-r--   1 ubuntu ubuntu  129M Dec 28 21:04 kube-apiserver.log
-rw-rw-r--   1 ubuntu ubuntu  1.8G Dec 28 20:58 kube-apiserver.log.1
-rw-rw-r--   1 ubuntu ubuntu  1.8G Dec 28 21:06 kube-apiserver.log.2
-rw-rw-r--   1 ubuntu ubuntu  2.5G Dec 28 21:04 kube-apiserver.log.3
-rw-rw-r--   1 ubuntu ubuntu  2.5G Dec 28 21:00 kube-apiserver.log.4
-rw-rw-r--   1 ubuntu ubuntu  2.5G Dec 28 21:02 kube-apiserver.log.5
-rw-rw-r--   1 ubuntu ubuntu   52M Dec 28 21:10 kube-controller-manager.log
-rw-rw-r--   1 ubuntu ubuntu  198M Dec 28 21:06 kube-controller-manager.log.1
-rw-rw-r--   1 ubuntu ubuntu  4.3G Dec 28 21:10 kube-controller-manager.log.2
-rw-rw-r--   1 ubuntu ubuntu  2.0M Dec 28 21:10 kube-proxy.log
-rw-rw-r--   1 ubuntu ubuntu   40M Dec 28 21:06 kube-scheduler.log
-rw-rw-r--   1 ubuntu ubuntu  618K Dec 28 20:56 kubelet.log
-rw-rw-r--   1 ubuntu ubuntu 1021K Dec 28 20:56 protokube.log
-rw-rw-r--   1 ubuntu ubuntu   39K Dec 28 21:10 sysctls
ubuntu@dev:/tmp/logs/artifacts/control-plane-us-east1-b-b1hb$ exit
```

File: tests/e2e/kubetest2-kops/deployer/dumplogs.go (hunk 5 above)
Comment (hakman) [reply tests/e2e/kubetest2-kops/deployer/dumplogs.go:55]: Please add this flag only for the scale tests. In other tests, we are fin with dumping more.

**Comment by upodroid** [comment]
This PR is ready to merged

@hakman

**Comment by upodroid** [comment]
bump

File: tests/e2e/scenarios/scalability/run-test.sh [hunk 6]
```diff
@@ -141,38 +131,38 @@ KUBETEST2_ARGS=()
...
+  KUBETEST2_ARGS+=("--boskos-resource-type=scalability-scale-project")
+  KUBETEST2_ARGS+=("--control-plane-instance-group-overrides=spec.rootVolume.type=pd-ssd")
+  KUBETEST2_ARGS+=("--max-nodes-to-dump=10")
+fi
+
 # More time for bigger clusters
 KUBETEST2_ARGS+=("--validation-wait=55m")
 KUBETEST2_ARGS+=("--validation-count=3")
-KUBETEST2_ARGS+=("--validation-interval=60s")
+KUBETEST2_ARGS+=("--validation-interval=70s")
```
Comment (hakman) [inline tests/e2e/scenarios/scalability/run-test.sh:147]: Why increase to 70s?

Comment (upodroid) [reply tests/e2e/scenarios/scalability/run-test.sh:147]: I did 70 first and then forgot to fix it when I rebased it.

Comment (hakman) [reply tests/e2e/scenarios/scalability/run-test.sh:147]: could you remove this part? I don't have any other objections on the merge.

File: tests/e2e/scenarios/scalability/run-test.sh [hunk 7]
```diff
@@ -53,30 +30,32 @@ if [[ -z "${K8S_VERSION:-}" ]]; then
   K8S_VERSION="$(curl -s -L https://dl.k8s.io/release/stable.txt)"
 fi
 
-# Download latest prebuilt kOps
-if [[ -z "${KOPS_BASE_URL:-}" ]]; then
-  KOPS_BASE_URL="$(curl -s https://storage.googleapis.com/kops-ci/bin/latest-ci.txt)"
-fi
-export KOPS_BASE_URL
+## FETCH MAHAMED's perf-tests fork
+git -C "${GOPATH}"/src/k8s.io/perf-tests remote add dev-fork https://github.com/borg-land/perf-tests.git
```
Comment (hakuna-matatah) [inline tests/e2e/scenarios/scalability/run-test.sh:34]: I think this `borg-land fork repo`  was used for testing purposes and forgot to change it back ?

File: tests/e2e/scenarios/scalability/run-test.sh [hunk 8]
```diff
@@ -86,12 +65,25 @@ echo "ADMIN_ACCESS=${ADMIN_ACCESS}"
...
+  create_args+=("--node-size=t3a.medium,t3.medium,t2.medium,t3a.large,c5a.large,t3.large,c5.large,m5a.large,m6a.large,m5.large,c4.large,c7a.large,r5a.large,r6a.large,m7a.large")
+  create_args+=("--node-volume-size=20")
+  create_args+=("--zones=us-east-2a,us-east-2b,us-east-2c")
+  create_args+=("--image=${INSTANCE_IMAGE:-ssm:/aws/service/canonical/ubuntu/server/20.04/stable/current/amd64/hvm/ebs-gp2/ami-id}")
+fi
+if [[ "${CLOUD_PROVIDER}" == "gce" ]]; then
+  create_args+=("--zones=us-east1-b,us-east1-c,us-east1-d")
+  create_args+=("--node-size=e2-standard-2")
+  create_args+=("--node-volume-size=30")
+  create_args+=("--master-volume-size=1000")
```
Comment (hakuna-matatah) [inline tests/e2e/scenarios/scalability/run-test.sh:79]: Interesting to see the volume size to be `1000`, out of curiosity, why would it need `1000` ?

File: tests/e2e/scenarios/scalability/run-test.sh [hunk 9]
```diff
@@ -182,29 +172,28 @@ if [[ "${CLOUD_PROVIDER}" == "aws" ]]; then
...
+  --create-args="${create_args[*]}" \
+  --test=clusterloader2 \
+  -- \
+  --provider="${CLOUD_PROVIDER}" \
+  --repo-root="${GOPATH}"/src/k8s.io/perf-tests \
+  --test-configs="${GOPATH}"/src/k8s.io/perf-tests/clusterloader2/testing/load/config.yaml \
+  --test-overrides="${GOPATH}"/src/k8s.io/perf-tests/clusterloader2/testing/load/overrides.yaml \
+  --extra-args="--experimental-prometheus-snapshot-to-report-dir=true" \
+  --kube-config="${HOME}/.kube/config"
+  # --test-overrides="${GOPATH}"/src/k8s.io/perf-tests/clusterloader2/testing/experiments/enable_restart_count_check.yaml \
```
Comment (hakuna-matatah) [inline tests/e2e/scenarios/scalability/run-test.sh:194]: are these intentionally commented out ? if so, why ?

**Comment by hakuna-matatah** [comment]
`Just` before this PR is about to be merged,  could we run `aws test at 5k scale presubmits ` to make sure everything is working as expected  @upodroid ?

File: tests/e2e/scenarios/scalability/run-test.sh (hunk 7 above)
Comment (upodroid) [reply tests/e2e/scenarios/scalability/run-test.sh]: Removed

File: tests/e2e/scenarios/scalability/run-test.sh (hunk 8 above)
Comment (upodroid) [reply tests/e2e/scenarios/scalability/run-test.sh:73]: Bigger disks have more iops on GCE

**Review by hakman — APPROVED** [review]

**Pull request merged**